### ISSUE TITLE
Problem with substituting sdata.table with external AnnData

### ISSUE THREAD

**Issue by gboscagli** [scverse/napari-spatialdata#293]
Hi all,
I'm starting to use `SpatialData `framework but I'm still missing something.
I'm working on this [public CosMx dataset](https://kero.hgc.jp/Breast_Cancer_Spatial.html). I'm able to read it as a `SpatialData` object and to visualize it using `Napari`, annotations included. However, I have done previous work on the same dataset and stored some annotations (louvain clusters, cell type annotations, etc...) in an `AnnData` object. At a first glance, I naively tried to substitute sdata.table with my `AnnData` (I saw you "recommended" it in scverse/spatialdata#270): it kinda works, except for the fact that when opened in `Napari` the table isn't recognized anymore in the "Tables annotating layer" menu, and thus I'm not able to plot any annotation. Maybe some references and metadata linking labels to annotations got lost during the transition: I thought of the missing `region`, `region_key` and `instance_key` metadata, but they are missing from the `SpatialData` table too. For the moment, I worked around it by concatenating specific columns from `AnnData` to `SpatialData` table, but I don't think this is the ideal solution. I also guess that adding my `AnnData` as additional table wouldn't solve my problem.

Do you have some recommendations for my issue?

**Comment by gboscagli** [comment]
Hi @LucaMarconato,
I used the following code to reproduce the error, don't know if it's what you meant me to do. It all worked well, even if I tried to `del sdata["table"]` and then set it again.

```
from spatialdata.datasets import blobs
sdata = blobs()

adata = sdata["table"]
sdata["table1"] = adata

interactive = Interactive(sdata)
interactive.run()
```

I think @melonora is aware of this, as we spoke about the same problem in an [Image.sc issue](https://forum.image.sc/t/avoid-labels-color-and-setting-directlabelcolormap-in-napari/100062). With my public `CosMx` dataset, I also tried manually adding annotation columns from external `AnnData` to the `sdata["table"]` I read from the raw folder: `Napari` isn't able to color just the plugged annotations, while with the native ones it works.

By now, I circumvented the problem by saving `AnnData obs` slot as the `metadata_file.csv`, and then read the folder as usual with `spatialdata_io.cosmx`. Napari now recognizes the colormap for all annotations, but I noticed another problem that I'm tracking in another issue so that I'm not making any more mess than I'm already making.

Could it depend on the dtype of the columns I'm trying to plot? They actually come from a `SpatialExperiment` object, converted into an `AnnData` object with an in-house function. I guess it's a very specific problem that could probably depend on my data, but maybe is not.

**Comment by LucaMarconato** [comment]
Yes in principle it could depend on the dtype.

> but I noticed another problem that I'm tracking in another issue so that I'm not making any more mess than I'm already making.

which is the other tracked issue? I could have a look if helpful.

Regarding the reproducible example, if I got it right you couldn't reproduce the bug with the `blobs` dataset, correct? Unfortunately without access to the data and without a reproducible example, investigating this bug becomes more time consuming and we are trying to prioritize other GitHub issues.

**Comment by gboscagli** [comment]
> which is the other tracked issue? I could have a look if helpful.

I haven't opened it yet. Basically, when I plot an obs annotation, a color is not "unique" for a certain discrete value. To be clearer, in the following image, pinkish cells can refer to several different annotations. Also, an annotation can be found with more than one color (i.e., a mural cell could be either pink or another color). I'm quite sure it is a Colormap problem, but I don't know how could be related to the original issue.

![image](https://github.com/user-attachments/assets/09863ed3-4e76-49b2-816f-6d63db58bc23)

> Unfortunately without access to the data and without a reproducible example, investigating this bug becomes more time consuming and we are trying to prioritize other GitHub issues.

The dataset is public (link above), I could just send my annotated `AnnData` object to be plugged and you can try to reproduce the bug. But I understand it's not as handy as the `blobs dataset` is.

**Comment by LucaMarconato** [comment]
> The dataset is public (link above), I could just send my annotated AnnData object to be plugged and you can try to reproduce the bug. But I understand it's not as handy as the blobs dataset is.

Sounds great! 

> (i.e., a mural cell could be either pink or another color). I'
Please check the code in this PR: https://github.com/scverse/napari-spatialdata/pull/304/files this is how one can assign custom colors (this is the same as what `scanpy` uses). I think that what you are experiencing is the result of a random colormap where some colors are too similar or collide.

**Comment by gboscagli** [comment]
> Sounds great!

Where do you prefer the .h5ad file to be sent?

> Please check the code in this PR: https://github.com/scverse/napari-spatialdata/pull/304/files this is how one can assign custom colors (this is the same as what scanpy uses). I think that what you are experiencing is the result of a random colormap where some colors are too similar or collide.

I will check the PR and keep you updated.

**Comment by LucaMarconato** [comment]
Via Zulip thanks: https://scverse.zulipchat.com/#user/480560

> I will check the PR and keep you updated.

sounds great, please if there are further comments on it let's follow up on a dedicated GitHub Issue.

**Comment by gboscagli** [comment]
**UPDATE**: I found out that the problem resided somewhere in my external `AnnData`. It was created from a `SpatialExperiment` object using `reticulate` in R with a converter in-house function that probably lost track of some metadata. Now I just transferred annotations as a `csv` file instead of a complete `AnnData`, and used it to read the object from scratch: when plugged inside a `SpatialData` object, `Napari` doesn't complain anymore and finally shows the colormap.

I'm closing this!
